apigee_edge.module

```diff
@@ -108,6 +114,41 @@ function apigee_edge_user_format_name_alter(&$name, AccountInterface $account) {
   }
 }
 
+/**
+ * Prepares variables for Apigee entity templates.
+ *
+ * Default template: apigee-entity.html.twig.
+ *
+ * @param array $variables
+ *   An associative array containing:
+ *   - elements: An array of elements to display in view mode.
+ */
+function template_preprocess_apigee_entity(array &$variables) {
+  $variables['view_mode'] = $variables['elements']['#view_mode'];
+  /** @var \Drupal\apigee_edge\Entity\EdgeEntityInterface $entity */
+  $entity = $variables['elements']['#entity'];
+
+  $variables['label'] = $entity->label();
+
+  if (!$entity->isNew() && $entity->hasLinkTemplate('canonical')) {
+    $variables['url'] = $entity->toUrl('canonical', ['language' => $entity->language()])->toString();
+  }
+
+  // Helpful $content variable for templates.
+  $variables += ['content' => []];
+  foreach (Element::children($variables['elements']) as $key) {
+    $variables['content'][$key] = $variables['elements'][$key];
+  }
+
+  // Helpful classes for front-end developers.
+  $variables['attributes']['class'][] = 'apigee-entity-default';
+  $variables['attributes']['class'][] = Html::getClass($entity->getEntityTypeId());
+  $variables['attributes']['class'][] = Html::getClass($entity->getEntityTypeId() . '--view-mode-' . $variables['elements']['#view_mode']);
```

Comment:
Although I understand the reasonds behind reference CR, but Drupal core and contrib modules can and do attach custom CSS classes to render arrays in preprocess functions so we can not just ignore them
![screeshot-2019-05-02-09-45-36](https://user-images.githubusercontent.com/1755573/57063349-37142380-6cc3-11e9-95a8-f3491d16d844.png)

https://github.com/drupal/core/blob/7d1adf98928d69d509b7b743f4abb01205813ba0/modules/contextual/contextual.module#L129

Until we have to support them we can also leverage them as well.

Reply:
> Until we have to support them we can also leverage them as well: https://github.com/apigee/apigee-edge-drupal/pull/181/files#diff-228def7411670fa59e3a8fc1a707afceR49

I think you misunderstand the features and how this works.

Sure, as long as we have preprocess (which may not actually be long at all TBH, as there are active issues trying to replace it), modules will be *able* to work with this variable.  However, this is *not* recommended as I mentioned, because it makes it more difficult to change things in the template. 

Preprocess is obviously still needed in some cases. The screenshot you posted is a prime example of where this is a good thing. Features like Contextual Links and Quick Edit are do not make sense to handle at template level, so using preprocess is perfectly sensible and acceptable because:

- They are used/needed across too many theme hooks / templates.
- You actually _want_ to make it hard for theme developers to remove those classes, because it will literally break functionality when if they do.

Our case here, working with an entity is very different.

```
{% set classes = attributes.class|merge([
  ...
]) %}
```

This is wrong.

1. `classes` in the Twig template are explicitly separate and printed via `attributes`.  This is so they can be easily tossed by the theme developer, if they're not wanted. 

2.  You don't want or need to do this `attributes.class|merge([`.  `attributes` has any classes and whatever other attributes exist in the object from preprocess or render elements.  Simply printing `attributes` and using `.addClass`, handles any merging that needs to occur.  You don't need to do anything to "support" this.  This is handled for ALL theme hooks by Drupal core uniformly in Drupal 8.   See:

    -  [\Drupal\Core\Template\Attribute](https://git.drupalcode.org/project/drupal/blob/8.7.x/core/lib/Drupal/Core/Template/Attribute.php)
    - [template_preprocess()](https://api.drupal.org/api/drupal/core%21includes%21theme.inc/function/template_preprocess/8.7.x)
    - [_template_preprocess_default_variables()](https://api.drupal.org/api/drupal/core%21includes%21theme.inc/function/_template_preprocess_default_variables/8.7.x)

HTH.

Reply:
It is weird, I do remember that without 
```
{% set classes = attributes.class|merge([
  ...
]) %}
```
The classes added by preprocessed did not appear before.